### PR TITLE
[node] Allow SharedArrayBuffer in Buffer.from

### PR DESCRIPTION
**Pull request by ThomasdenH**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/nodejs/node/commit/2a2ec9dbc3375cd1b8be58fb2ed522247474bc20
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Changes `Buffer.from` and `new Buffer` to accept a `SharedArrayBuffer`. Essentially identical to #19656.

### CODE REVIEW

**Comment by ThomasdenH** [comment]
The errors are not in the code changed by this PR.

File: types/node/node-tests.ts [hunk 1]
```diff
@@ -35,6 +35,11 @@ import Module = require("module");
 // Specifically test buffer module regression.
 import { Buffer as ImportedBuffer, SlowBuffer as ImportedSlowBuffer } from "buffer";
 
+// Forward-declare SharedArrayBuffer (available starting es2017)
+interface SharedArrayBuffer {
```
Comment (Flarna) [inline types/node/node-tests.ts]: This shouldn't be needed here as the forward declaration is already in index.d.ts.

File: types/node/index.d.ts [hunk 2]
```diff
@@ -26,12 +26,18 @@
...
 //                 Zane Hannan AU <https://github.com/ZaneHannanAU>
+//                 Thomas den Hollander <https://github.com/ThomasdenH>
 //                 Eugene Y. Q. Shen <https://github.com/eyqs>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
 /** inspector module types */
 /// <reference path="./inspector.d.ts" />
 
+// Forward-declare SharedArrayBuffer (available starting es2017)
+interface SharedArrayBuffer {
```
Comment (Flarna) [inline types/node/index.d.ts]: Please move this down to other forward declarations (`Iterable<T>`,...).

File: types/node/index.d.ts [hunk 3]
```diff
@@ -26,12 +26,18 @@
...
+    readonly byteLength: number;
```
Comment (Flarna) [inline types/node/index.d.ts]: This somehow reduces the type safety as everything having a `readonly byteLength: number;` matches here. Should we add more properties?
I'm not sure here.

**Review by Flarna — CHANGES_REQUESTED** [review]
Consider to adapt also `Buffer.byteLength()`

File: types/node/index.d.ts (hunk 3 above)
Comment (ThomasdenH) [reply types/node/index.d.ts]: The only other property is the slice method, I have added it.

File: types/node/index.d.ts (hunk 2 above)
Comment (ThomasdenH) [reply types/node/index.d.ts]: Done.

File: types/node/node-tests.ts (hunk 1 above)
Comment (ThomasdenH) [reply types/node/node-tests.ts]: Done.

File: types/node/index.d.ts (hunk 3 above)
Comment (Flarna) [reply types/node/index.d.ts]: This change caused one of your tests to fail.

**Comment by ThomasdenH** [comment]
I changed `Buffer.byteLength()` as well.

File: types/node/index.d.ts (hunk 3 above)
Comment (ThomasdenH) [reply types/node/index.d.ts]: Of course.

**Review by Flarna — APPROVED** [review]

**Comment by ThomasdenH** [comment]
The errors are not in node but in other packages.

**Comment by Flarna** [comment]
@andy-ms @rbuckton CI fails in puppeteer are unrelated like in other node PRs. Would be nice if you could also take a look in this regarding the forward declaration of `SharedArrayBuffer`.

**Pull request merged**